doc/compiler/control/CompilationControl.md

```diff
@@ -0,0 +1,304 @@
+compiled body is stored in the Data Cache. Space in the Code and Data
+Caches have to be allocated on an as-needed basis. Additionally,
+these allocations have to be synchronized since multiple Compilation
+Threads can be active simultaneously. As such, each Compilation Thread,
+as part of the compilation, will reserve a Code Cache to ensure that
+no other thread can write to it.
+
+There unfortunately exists a bit of name overloading. When speaking
+generally about a place to store compiled bodies, the term "Code Cache"
+is used. However, the actual chunk of memory is known as the "Code
+Respository", and individual subsections of the Code Repository are
+called "Code Caches". By default, the JIT allocates a 256 MiB Code
+Repository. This is then carved into 2 MiB Code Caches. It is this
+2 MiB subsection of the Code Reponsitory that a Compilation Thread
+reserves during a compilation.
+
+### Trampolines
+
+A trampolines is essentially a snippet of code that is used branch to
+a location far from the current Instruction Pointer. Each Code Cache
+has a section for storing Trampolines. This is so that methods in a
+Code Cache can share Trampolines. The Code Caches grows towards
+increasing addresses, while the Trampoline section grows towards
+decreasing addresses.
+
+### Code Cache Reclamation
+
+When a method is recompiled, the old body eventually stops being used,
+which is a waste of space in the Code Cache. In order to maximize the
+usable space in the Code Cache, the JIT reclaims stale bodies; see the
+[Code Cache Reclamation doc](../runtime/CodeCacheReclamation.md) for
+more details.
+
+
+# Influencing Factors
+
+There are many factors that influence the decisions the Control
+Infrastructure makes. The JIT has to balance improving application
+performance with not taking away too many resources from the
+application. The following are some of the considerations that have
+the biggest impact and influence.
+
+## JIT State
+
+The JIT maintains several states that reflect the various execution
+characteristics of an application. The Compilation Control heuristics
+depend on the state, or phase, the JIT is in. The JIT transitions
+between the following states:
+
+- IDLE_STATE
+  - Decisions focus on minimizing Compilation Thread CPU usage.
+- STARTUP_STATE
+  - Decisions focus on making Startup as fast as possible.
+- RAMPUP_STATE
+  - Decisions focus on making Rampup as fast as possible.
+- STEADY_STATE
+  - Decisions focus on maintaining throughput.
+- DEEPSTEADY_STATE
+  - Decisions focus on maintaining throughput while minimizing CPU
+    consumed by the JIT (including the Sampler Thread).
+
+## Available Resources
+
+The JIT should minimize its impact on application performance.
+Therefore, how aggressively it can compile methods depends on
+the resources available to the JVM. Multiple compilation threads may
+only be activated if the system is not resource constrained.
+To minimize memory usage, only one higher optimization level
```

Comment:
I vaguely remember that I performed some changes in this area to help Spark rampup faster. I don't recall what I did though.